### PR TITLE
AArch64 portability fixes

### PR DESCRIPTION
**Pull request by akopytov**
A set of fixes to make PerconaFT build and run on the AArch64 (64-bit ARMv8) architecture.

### CODE REVIEW

**Comment by prohaska7** [comment]
Did you consider using clock_gettime CLOCK_REALTIME instead of the various low level processor instructions?  The realtime clock on linux is very fast since it memory maps the clock info into the user process space (see vdso).

**Comment by akopytov** [comment]
Hi Rich,

I didn't, I was just adding a platform-specific implementation for toku_time_now() which uses RDTSC on x86.

But I see a couple of issues with clock_gettime(CLOCK_REALTIME):

- VDSO is not always available (it isn't on my hardware)
- nothing really beats a single CPU instruction in terms of overhead, and comments for toku_time_now() suggest it is supposed to be a low-overhead timer.

I did not measure overhead for instructions reading CPU cycles information vs. clock_gettime(). However, PERFORMANCE_SCHEMA uses the 'cycles' timer in contexts where it needs the lowest overhead timer and falls back to the nanoseconds (usually clock_gettime()-based) timer when the cycles one is not available.

"Additional Comments" in mysys/my_rdtsc.c is also an interesting read.

**Comment by prohaska7** [comment]
What hardware are you running on?  I thought that vdso was a software feature in linux that makes gettimeofday (and other clock related system calls) fast.  It does this my avoid kernel system calls.

**Comment by akopytov** [comment]
Sorry, I spoke too soon. VDSO is available on my platform, but whether it is actually used to accelerate each particular syscall depends on glibc, and that itself depends on hardware capabilities. I've seen certain system calls like sched_getcpu() using VDSO on x86, but not on my ARM-based platform.

It's hard to figure out from the glibc code whether VDSO is actually used for clock_gettime() in my case.

I just decided to compare overhead of a RDTSC-like timer and clock_gettime(), but noticed that code in mysys/my_rdtsc.c already does that, and that information is exposed through PERFORMANCE_SCHEMA. This is what I got:
```
mysql> select * from performance_timers;
+-------------+-----------------+------------------+----------------+
| TIMER_NAME  | TIMER_FREQUENCY | TIMER_RESOLUTION | TIMER_OVERHEAD |
+-------------+-----------------+------------------+----------------+
| CYCLE       |        99935323 |                1 |              0 |
| NANOSECOND  |      1000000000 |                1 |              5 |
| MICROSECOND |         1000000 |                1 |              6 |
| MILLISECOND |            1037 |                1 |              6 |
| TICK        |             103 |                1 |             58 |
+-------------+-----------------+------------------+----------------+
5 rows in set (0.00 sec)
```

If I read the code in my_rdtsc.c correctly, the above means that clock_gettime() is about 5 'cycle' units more expensive than the CYCLE timer itself. A 'cycle' in my case has nothing to do with CPU cycles, but is about 10 nanoseconds based on frequency, which means clock_gettime() has about 50 nanoseconds higher overhead as compared to the RDTSC-like timer.

**Comment by georgelorchpercona** [comment]
Hey Alexey,
I'd be glad to test and merge your changes, but I need you to alter your commit message to include our GPL-2 license release as mentioned here https://github.com/percona/PerconaFT#contributing

Thanks!

**Comment by akopytov** [comment]
Hi George,

That link says I need to contribute under the BSD 2-Clause license, not GPL-2. Also, do I really need to include the license text into each of the 5 commits, or just changing the pull request comment for all commits is enough?

**Comment by georgelorchpercona** [comment]
Good question, the lawyers had us change the contribution requirement to the BSD 2-Clause as defined in the template here https://opensource.org/licenses/BSD-2-Clause/

I haven't reviewed your contribution yet and didn't notice the 5 different commits and discreet fixes. We don't get a lot of contributions so are figuring this out as they come. Usually I would ask that the commits get squashed, but these all look like unrelated changes towards your goal of AArch64 compilation, which makes it reasonable to keep them apart.

In this case, I believe you have two choices, one is to go back and change all of the commit messages, the other and easier choice is to put the license text into the initial pull request description/comment that covers all 5 commits.

**Comment by akopytov** [comment]
Thanks for clarifications. I updated all 5 commits with the BSD 2-Clause license text.

File: portability/toku_time.h [hunk 1]
```diff
@@ -98,9 +98,17 @@ double tokutime_to_seconds(tokutime_t)  __attribute__((__visibility__("default")
 
 // Get the value of tokutime for right now.  We want this to be fast, so we expose the implementation as RDTSC.
 static inline tokutime_t toku_time_now(void) {
+#if defined(x86_64) || defined(__i386__)
```
Comment (georgelorchpercona) [inline portability/toku_time.h]: Fails to build on all supported intel platforms (all 64 bit), this should probably be __x86_64__

```22:36:07 /mnt/workspace/PerconaFT-param/BUILD_TYPE/release/Host/ubuntu-precise-64bit/portability/toku_time.h:110:2: error: #error No timer implementation for this platform
22:36:07  #error No timer implementation for this platform
22:36:07   ^
22:36:07 /mnt/workspace/PerconaFT-param/BUILD_TYPE/release/Host/ubuntu-precise-64bit/portability/toku_time.h: In function ‘tokutime_t toku_time_now()’:
22:36:07 /mnt/workspace/PerconaFT-param/BUILD_TYPE/release/Host/ubuntu-precise-64bit/portability/toku_time.h:112:1: error: no return statement in function returning non-void [-Werror=return-type]
22:36:07  }
22:36:07  ^
22:36:07 cc1plus: all warnings being treated as errors```

see http://jenkins.percona.com/view/TokuDB/job/PerconaFT-param/54/

**Comment by akopytov** [comment]
Fixed x86_64 builds by replacing `x86_64` with `__x86_64__` in `#ifdef`.

File: portability/toku_time.h (hunk 1 above)
Comment (akopytov) [inline portability/toku_time.h]: Fixed. Thanks.

**Comment by georgelorchpercona** [comment]
Thanks Alexey, this is testing much better.

**Comment by georgelorchpercona** [comment]
http://jenkins.percona.com/view/TokuDB/job/PerconaFT-param/55/

**Comment by akopytov** [comment]
I'm not sure I read the Jenkins status above correctly. Is it good to merge, or some action required on my end?

**Comment by georgelorchpercona** [comment]
Sorry Alexey, yes, that is the ctest suite which has recently become unstable due to some changes in jenkins. I also have pending PS jobs that will get logged here today. So far everything looks fine and no more is required from you.

**Comment by georgelorchpercona** [comment]
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1211/
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/225/

**Comment by akopytov** [comment]
Thanks, George!

**Pull request merged**